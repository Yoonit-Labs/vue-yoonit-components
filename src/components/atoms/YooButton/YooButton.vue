<template lang="pug">
.yoo__container--button
  button(
    :class="['yoo-btn', ...takeModifier]"
    :disabled="disabled"
    @click="$emit('onClick')"
  )
    i(
      v-if="icon.length && iconPosition === 'left'"
      :class="['yoo-btn__icon', takeIconStyle, takeIconName, ...takeModifierIcon]"
      :style="takeIconColor"
    )
    p.yoo-btn__text {{ text }}

    i(
      v-if="icon.length && iconPosition === 'right'"
      :class="['yoo-btn__icon', takeIconStyle, takeIconName, ...takeModifierIcon]"
      :style="takeIconColor"
    )
</template>
<script>
import PropsConfig from '@/components/atoms/YooButton/YooButton.config'
export default {
  name: 'YooButton',
  props: {
    variation: {
      type: String,
      default: 'base',
      validator: value => PropsConfig.variation.options.includes(value)
    },
    buttonSize: {
      type: String,
      default: 'normal',
      validator: value => PropsConfig.buttonSize.options.includes(value)
    },
    fill: {
      type: String,
      default: 'primary',
      validator: value => PropsConfig.fill.options.includes(value)
    },
    icon: {
      type: String,
      default: ''
    },
    iconPosition: {
      type: String,
      default: 'left',
      validator: value => PropsConfig.iconPosition.options.includes(value)
    },
    iconSize: {
      type: String,
      default: 'normal',
      validator: value => PropsConfig.iconSize.options.includes(value)
    },
    iconStyle: {
      type: String,
      default: 'solid',
      validator: value => PropsConfig.iconStyle.options.includes(value)
    },
    animation: {
      type: String,
      default: 'none',
      validator: value => PropsConfig.animation.options.includes(value)
    },
    disabled: {
      type: Boolean,
      default: false
    },
    text: {
      type: String,
      default: ''
    },
    hover: {
      type: Boolean,
      default: false
    },
    active: {
      type: Boolean,
      default: false
    },
    iconCustomColor: {
      type: String,
      default: '',
      validator (value) {
        if (value.includes('#')) {
          const rx = /^#([0-9a-f]{6}|[0-9a-f]{3})$/i
          return value.match(rx)
        } else {
          return value === ''
        }
      }
    }
  },
  computed: {
    takeIconColor () {
      return { color: `${this.iconCustomColor}` }
    },
    takeModifierIcon () {
      const block = 'yoo-btn__icon'
      const classList = []
      classList.push(`${block}--${this.iconSize}`)
      if (this.iconPosition && this.text) {
        classList.push(`${block}--${this.iconPosition}`)
      }
      return classList
    },
    takeModifier () {
      const block = 'yoo-btn'
      const classList = []
      classList.push(`${block}__fill--${this.fill}`)
      classList.push(`${block}--${this.variation}`)
      classList.push(`${block}--${this.buttonSize}`)
      classList.push(`${block}--animation-${this.animation}`)
      if (this.fill && this.hover && !this.disabled) {
        classList.push(`${block}--hover`)
      }
      if (this.fill && this.active && !this.disabled) {
        classList.push(`${block}--active`)
      }
      if (this.fill && this.disabled) {
        classList.push(`${block}--disabled`)
      }
      return classList
    },
    takeIconName () {
      return `fa-${this.icon}`
    },
    takeIconStyle () {
      return this.iconStyle === 'solid'
        ? 'fas' // fontawesome solid
        : 'far' // fontawesome regular
    }
  }
}
</script>

<style src="./YooButton.sass" lang="sass" scoped></style>
