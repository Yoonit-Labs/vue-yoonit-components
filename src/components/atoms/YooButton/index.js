import yooButton from '@/components/atoms/YooButton/YooButton.vue'
export default yooButton
