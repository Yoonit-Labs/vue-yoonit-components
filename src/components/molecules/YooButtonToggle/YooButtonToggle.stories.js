import yooButtonToggle from './YooButtonToggle.vue'
import PropsConfig from './YooButtonToggle.config'

export default {
  title: 'Yoonit/YooButtonToggle',
  component: yooButtonToggle,
  argTypes: {
    size: {
      control: {
        type: 'select',
        options: PropsConfig.size.options
      }
    }
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooButtonToggle },
  template:
    '<yoo-button-toggle v-bind="$props"></yoo-button-toggle>'
})

export const Custom = Template.bind({})
Custom.args = { options: ['Yoo Button Toggle', 'Yoo Button Toggle 1'] }
