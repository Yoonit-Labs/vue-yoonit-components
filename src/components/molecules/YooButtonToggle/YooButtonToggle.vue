<template lang="pug">
  YooFlexLayout(
    flexDirection="column",
    flexWrap="nowrap"
    class="yoo-button-toggle"
  )
    YooFlexLayout(
      flexDirection="row",
      alignContent="center",
      alignItems="center",
      justifyContent="center",
      flexWrap="nowrap"
      class="yoo-button-toggle__container"
    )
      YooButton(
        v-for="(option, index) in options",
        :key="option.toLowerCase()",
        v-show="options[index]"
        :text="option",
        :variation="selectedIndex === index ? 'base' : 'clear'",
        :fill="selectedIndex === index ? 'primary' : 'default'",
        size="normal",
        style="width: 100%;"
        @onClick="doEmitChoice($event, index)"
      )
</template>

<script>
import YooButton from '@/components/atoms/YooButton/YooButton.vue'

import YooFlexLayout from '@/components/quarks/YooFlexLayout/YooFlexLayout.vue'

export default {
  name: 'YooButtonToggle',
  props: {
    options: {
      type: Array,
      required: true
    },
    current: {
      type: Number,
      default: 0
    },
    toggleInit: {
      type: Number
    }
  },
  components: {
    YooFlexLayout,
    YooButton
  },
  data: () => ({
    selectedIndex: 0
  }),
  mounted () {
    this.selectedIndex = this.current
  },
  methods: {
    doEmitChoice (event, index) {
      this.selectedIndex = index
      this.$emit('tapChoice', index)
    }
  }
}
</script>

<style src="./YooButtonToggle.sass" lang="sass" scoped></style>
