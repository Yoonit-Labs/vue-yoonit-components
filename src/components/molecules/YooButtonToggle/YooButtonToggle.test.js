import { shallowMount } from '@vue/test-utils'
import YooButtonToggle from '@/components/molecules/YooButtonToggle/YooButtonToggle.vue'
import YooButton from '@/components/atoms/Button/Button.vue'

const SlotText = 'Default Slot Text'

const mountComponent = () => {
  return shallowMount(YooButtonToggle, {
    slots: { default: SlotText },
    propsData: {
      options: ['Yoo Button Toggle', 'Yoo Button Toggle 1']
    }
  })
}

describe('YooButtonToggle Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountComponent()
  })

  it('Matches Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', async () => {
    expect(wrapper.find('.yoo-button-toggle').exists()).toBe(true)
    expect(wrapper.find('.yoo-button-toggle__container').exists()).toBe(true)
    // Test check YooButton
    wrapper = shallowMount(YooButton)
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.yoo-btn').exists()).toBe(true)
  })

  describe('Events', () => {
    describe('doEmitChoice', () => {
      it('Emits doEmitChoice Event', async () => {
        // Test check YooButton
        wrapper = shallowMount(YooButton, { propsData: { fill: 'primary' } })
        await wrapper.vm.$nextTick()
        wrapper.find('.yoo-btn').trigger('click')
        expect(wrapper.emitted()).toHaveProperty('onClick')
      })
    })
  }) // describe Events
})
