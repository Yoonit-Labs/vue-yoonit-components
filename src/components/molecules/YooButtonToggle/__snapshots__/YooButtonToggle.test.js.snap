// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooButtonToggle Component Matches Snapshot 1`] = `
<div class="yoo-flex-layout yoo-button-toggle" style="flex-direction: column; flex-wrap: nowrap; justify-content: center; align-items: stretch; align-content: stretch;">
  <div class="yoo-flex-layout yoo-button-toggle__container" style="flex-direction: row; flex-wrap: nowrap; justify-content: center; align-items: center; align-content: center;">
    <div class="yoo__container--button" size="normal" style="width: 100%;"><button class="yoo-btn yoo-btn__fill--primary yoo-btn--base yoo-btn--normal yoo-btn--animation-none">
        <!---->
        <p class="yoo-btn__text">Yoo Button Toggle</p>
        <!---->
      </button></div>
    <div class="yoo__container--button" size="normal" style="width: 100%;"><button class="yoo-btn yoo-btn__fill--default yoo-btn--clear yoo-btn--normal yoo-btn--animation-none">
        <!---->
        <p class="yoo-btn__text">Yoo Button Toggle 1</p>
        <!---->
      </button></div>
  </div>
</div>
`;
